lib/kubeclient/informer.rb

```diff
@@ -0,0 +1,85 @@
+module Kubeclient
+  # caches results for multiple consumers to share and keeps them updated with a watch
+  class Informer
+    def initialize(client, resource_name, reconcile_timeout: 15 * 60)
+      @client = client
+      @resource_name = resource_name
+      @reconcile_timeout = reconcile_timeout
+      @cache = nil
+      @started = nil
+      @watching = []
+    end
+
+    def list
+      @cache.values
+    end
+
+    def watch(&block)
+      with_watching(&block)
+    end
+
+    def watch_diff(&block)
+      with_watching(&block)
+    end
+
+    # not implicit so users know they have to `stop`
+    def start_worker
+      @worker = Thread.new do
+        loop do
+          fill_cache
+          watch_to_update_cache
+        rescue StandardError
+          # need to keep retrying since we work in the background
+        ensure
+          sleep(1) # do not overwhelm the api-server if we are somehow broken
+        end
+      end
+      sleep(0.01) until @cache
+    end
+
+    def stop_worker
+      @worker&.kill # TODO: be nicer ?
+    end
+
+    private
+
+    def with_watching
+      queue = Queue.new
+      @watching << queue
+      loop do
+        x = queue.pop
+        yield(x)
+      end
+    ensure
+      @watching.delete(queue)
+    end
+
+    def cache_key(resource)
+      resource.dig(:metadata, :selfLink)
+    end
+
+    def fill_cache
+      reply = @client.get_entities(nil, @resource_name, raw: true, resource_version: '0')
+      @cache = reply[:items].each_with_object({}) do |item, h|
+        h[cache_key(item)] = item
+      end
+      @started = reply.dig(:metadata, :resourceVersion)
+    end
+
+    def watch_to_update_cache
+      Timeout.timeout(@reconcile_timeout) do
+        @client.watch_entities(@resource_name, watch: true, resource_version: @started) do |notice|
+          case notice[:type]
+          when 'ADDED', 'MODIFIED' then @cache[cache_key(notice[:object])] = notice[:object]
+          when 'DELETED' then @cache.delete(cache_key(notice[:object]))
+          when 'ERROR' then break # restart
+          else raise "Unsupported event type #{notice[:type]}"
```

Comment:
If this happens (shouldn't), it'd raise in the worker thread.  Would ruby even print it?

Reply:
might be most consistent to ignore it so the worker does not crash ...
signaling back is kinda tricky, that's what I'm planing to use a logger for